### PR TITLE
Add fullscreen API

### PR DESCRIPTION
**Pull request by MortimerGoro**
- Receive DOM fullscreen state via wpe_view_backend_set_fullscreen
- Dispatch exit/enter fullscreem from libwpe to the client (e.g. WPEView in Webkit)

----

Corresponding WPE WebKit patch: https://bugs.webkit.org/show_bug.cgi?id=227951

### CODE REVIEW

File: src/view-backend.c [hunk 1]
```diff
@@ -187,3 +197,31 @@ wpe_view_backend_dispatch_touch_event(struct wpe_view_backend* backend, struct w
     if (backend->input_client)
         backend->input_client->handle_touch_event(backend->input_client_data, event);
 }
+
+void
+wpe_view_backend_set_fullscreen_handler(struct wpe_view_backend* backend, void (*handler)(void*, bool), void* data)
```
Comment (aperezdc) [inline src/view-backend.c:202]: Same here: is it expected that this function would be called only once per view backend? Do we want to allow changing the handler?

We may want to cover the case in which fullscreening is requested by the application using WPE determines that it is not currently possible. I think the callback should have type `bool (*))(void*, bool)`; and if it returns `false`, cancel the fullscreening operation on the WebKit side.

Finally, it is unclear who has to implement a “client” and who has to provide a “handler”, having some documentation about what each does would bee great.

**Review by aperezdc — COMMENTED** [review]
Functionality-wise I think the patch looks fine, though it was a bit hard to follow what's going on and how the API is to be used without looking at the accompanying [patch for WebKit](https://bugs.webkit.org/show_bug.cgi?id=227951) and the [PR for Cog](https://github.com/Igalia/cog/pull/329). Would it be possible to add some reference documentation?

File: include/wpe/view-backend.h [hunk 2]
```diff
@@ -179,6 +189,28 @@ WPE_EXPORT
 void
 wpe_view_backend_dispatch_touch_event(struct wpe_view_backend*, struct wpe_input_touch_event*);
 
+struct wpe_view_backend_fullscreen_client {
+    void (*request_enter_fullscreen)(void*);
+    void (*request_exit_fullscreen)(void*);
+
+
+    /*< private >*/
+    void (*_wpe_reserved0)(void);
```
[the remark this answers is not in the document]
Comment (MortimerGoro) [reply include/wpe/view-backend.h:205]: done

File: src/view-backend.c (hunk 1 above)
Comment (MortimerGoro) [reply src/view-backend.c]: I have improved the API with better function names, some comments and implementing the wait for the new exported image we talked about

File: src/view-backend.c [hunk 3]
```diff
@@ -93,6 +96,13 @@ wpe_view_backend_set_input_client(struct wpe_view_backend* backend, const struct
     backend->input_client_data = client_data;
 }
 
+void
+wpe_view_backend_set_fullscreen_client(struct wpe_view_backend* backend, const struct wpe_view_backend_fullscreen_client* client, void* client_data)
+{
+    backend->fullscreen_client = client;
+    backend->fullscreen_client_data = client_data;
```
[the remark this answers is not in the document]
Comment (MortimerGoro) [reply src/view-backend.c:103]: Do you have any example on how to assert that check? I mean do you mean adding a assert/panic or similar.

Checking a similar function (wpe_view_backend_set_input_client) I don't see it having the check or the destroy_notify callback

[the remark this answers is not in the document]
Comment (aperezdc) [reply src/view-backend.c:103]: We don't have such checks in other functions, but it would be good to have them. I think a plain `assert(!backend->fullscreen_client)` should be enough here 🤔

File: src/view-backend.c [hunk 4]
```diff
@@ -65,6 +65,9 @@ wpe_view_backend_destroy(struct wpe_view_backend* backend)
     backend->input_client = 0;
     backend->input_client_data = 0;
 
+    backend->fullscreen_client = 0;
+    backend->fullscreen_client_data = 0;
```
Comment (aperezdc) [inline src/view-backend.c:69]: I'm somewhat amused that this does not cause compiler warnings; we should be using `NULL` here because the members are pointers 🤔. Anyhoo, let's use it:

```suggestion
    backend->fullscreen_client = NULL;
    backend->fullscreen_client_data = NULL;
```

File: src/view-backend.c (hunk 3 above)
[the remark this answers is not in the document]
Comment (aperezdc) [reply src/view-backend.c:103]: I think you might have forgotten to add the assertion here 😉

File: include/wpe/view-backend.h [hunk 5]
```diff
@@ -97,6 +99,16 @@ WPE_EXPORT
...
 
+/* Sets the DOM client that listend to fullscreen requests and state changes from platform */
+WPE_EXPORT
+void
+wpe_view_backend_set_dom_fullscreen_client(struct wpe_view_backend*, const struct wpe_view_backend_fullscreen_client*, void*);
+
+/* Sets the platform function that handles fullscreen requests from DOM */
+WPE_EXPORT
+void
+wpe_view_backend_set_platform_fullscreen_handler(struct wpe_view_backend*, bool (*handler)(void*, bool), void*);
```
Comment (aperezdc) [inline include/wpe/view-backend.h:110]: Could we have a `typedef` for the callback function? I think it will be much more readable and would allow documenting it. We could add something like this, alongside the type definitions in the beginning of the file:

```c
/**
 * wpe_view_backend_fullscreen_callback:
 * @userdata: ...
 * @enable: ...
 * 
 * Type of functions used by an embedder to implement fullscreening web views.
 *
 * Returns: Whether the operation was completed.
 */
typedef bool (*wpe_view_backend_fullscreen_callback)(void *userdata, bool enable);
```

and then here, change the declaration of the function to:

```suggestion
wpe_view_backend_set_platform_fullscreen_handler(struct wpe_view_backend*,
        wpe_view_backend_fullscreen_callback callback, void* userdata);
```

**Review by aperezdc — CHANGES_REQUESTED** [review]
This is starting to look ready for merging, thanks for working on it! 💪🏼 

I have left a few comments with suggestions, and it would be great if we could have @zdobersek and/or @carlosgcampos check this and rubber-stamp that they are okay with the new API.

File: include/wpe/view-backend.h [hunk 6]
```diff
@@ -97,6 +99,16 @@ WPE_EXPORT
 void
 wpe_view_backend_set_input_client(struct wpe_view_backend*, const struct wpe_view_backend_input_client*, void*);
 
+/* Sets the DOM client that listend to fullscreen requests and state changes from platform */
```
[the remark this answers is not in the document]
Comment (MortimerGoro) [reply include/wpe/view-backend.h]: Done, thanks for the documentation example!

File: include/wpe/view-backend.h [hunk 7]
```diff
@@ -179,6 +191,50 @@ WPE_EXPORT
 void
 wpe_view_backend_dispatch_touch_event(struct wpe_view_backend*, struct wpe_input_touch_event*);
 
+struct wpe_view_backend_fullscreen_client {
```
[the remark this answers is not in the document]
Comment (MortimerGoro) [reply include/wpe/view-backend.h:243]: Done, thanks for the documentation example!

File: include/wpe/view-backend.h [hunk 8]
```diff
@@ -179,6 +191,50 @@ WPE_EXPORT
...
+void
+wpe_view_backend_dispatch_did_exit_fullscreen(struct wpe_view_backend*);
+
+WPE_EXPORT
+void
+wpe_view_backend_dispatch_request_enter_fullscreen(struct wpe_view_backend*);
+
+WPE_EXPORT
+void
+wpe_view_backend_dispatch_request_exit_fullscreen(struct wpe_view_backend*);
```
[the remark this answers is not in the document]
Comment (MortimerGoro) [reply include/wpe/view-backend.h]: Done

File: src/view-backend.c (hunk 4 above)
Comment (MortimerGoro) [reply src/view-backend.c]: Done. I normally use NULL (or nullptr in C++11) but followed the style of the other setters which were using 0.

File: include/wpe/view-backend.h (hunk 5 above)
Comment (MortimerGoro) [reply include/wpe/view-backend.h]: Done, thanks for the example!

**Review by aperezdc — CHANGES_REQUESTED** [review]
This looks almost ready, thanks! It will only need a couple of touch ups in the documentation comments.

In the meantime, I will try to get someone else to do a review as well, I would prefer to have one more person agree on new API before merging.

As a side note, I noticed a bug in HotDoc while trying to build the documentation added by this MR locally, and I have reported it: https://github.com/hotdoc/hotdoc/issues/230 🙃

File: include/wpe/view-backend.h [hunk 9]
```diff
@@ -97,6 +99,51 @@ WPE_EXPORT
...
+ * Since: 1.12
+ */
+WPE_EXPORT
+void
+wpe_view_backend_set_dom_fullscreen_client(struct wpe_view_backend*, const struct wpe_view_backend_fullscreen_client*, void*);
+
+/**
+ * wpe_view_backend_fullscreen_handler:
+ * @userdata: ...
+ * @enable: ...
```
[the remark this answers is not in the document]
Comment (MortimerGoro) [reply include/wpe/view-backend.h]: Done. Your explanations were great, I missed filling that in this comment :)

File: include/wpe/view-backend.h [hunk 10]
```diff
@@ -97,6 +99,51 @@ WPE_EXPORT
...
+
+/**
+ * wpe_view_backend_set_platform_fullscreen_handler:
+ * @view_backend: (transfer none): The view backend to obtains events from.
+ * @handler: (transfer none): Function used by an embedder to implement fullscreening web views.
+ * @userdata: (transfer none): User data passed to the handler function.
+ * 
+ * Handler function set by an embedder to implement fullscreening web views.
+ * 
+ * This function must be only used once for a given @view_backend, the client
```
[the remark this answers is not in the document]
Comment (MortimerGoro) [reply include/wpe/view-backend.h]: done

File: include/wpe/view-backend.h [hunk 11]
```diff
@@ -179,6 +226,98 @@ WPE_EXPORT
 void
 wpe_view_backend_dispatch_touch_event(struct wpe_view_backend*, struct wpe_input_touch_event*);
 
+/**
+ * wpe_view_backend_fullscreen_client:
+ * @did_enter_fullscreen: Invoked after fullscreen has been successfully entered.
+ * @did_exit_fullscreen: Invoked after fullscreen has been exited.
+ * @request_enter_fullscreen: Invoked after user has requested to enter fullscreen.
+ * @request_exit_fullscreen: Invoked after user has requested to exit fullscreen.
+ * @...
```
[the remark this answers is not in the document]
Comment (MortimerGoro) [reply include/wpe/view-backend.h]: Done

**Review by aperezdc — COMMENTED** [review]
This looks mergeable to me. Waiting to have a second opinion on the new API, hopefully @zdobersek and/or @carlosgcampos can take a look soon 👍🏼

File: include/wpe/view-backend.h [hunk 12]
```diff
@@ -97,6 +99,53 @@ WPE_EXPORT
...
+ * Handler function set by an embedder to implement fullscreening web views.
+ * 
+ * This function must be only used once for a given @view_backend, the handler
+ * cannot be changed once it has been set.
+ *
+ * Since: 1.12
+ */
+WPE_EXPORT
+void
+wpe_view_backend_set_platform_fullscreen_handler(struct wpe_view_backend*, wpe_view_backend_fullscreen_handler handler, void* userdata);
```
Comment (carlosgcampos) [inline include/wpe/view-backend.h:147]: Same for platform here, just set_fullscreen_handler

File: include/wpe/view-backend.h [hunk 13]
```diff
@@ -97,6 +99,53 @@ WPE_EXPORT
 void
 wpe_view_backend_set_input_client(struct wpe_view_backend*, const struct wpe_view_backend_input_client*, void*);
 
+/**
+ * wpe_view_backend_set_dom_fullscreen_client:
```
[the remark this answers is not in the document]
Comment (MortimerGoro) [reply include/wpe/view-backend.h]: Thanks for the review!

Done. I agree with the change, the new documentation already specifies what it can be used for.

File: include/wpe/view-backend.h (hunk 12 above)
Comment (MortimerGoro) [reply include/wpe/view-backend.h]: Done

**Review by aperezdc — APPROVED** [review]
Thanks @carlosgcampos for commenting, I think we can consider this now approved, right?

**Comment by aperezdc** [comment]
Let's merge this, I would like to make a development release including this feature, and it looks like @carlosgcampos should be happy now with the function name changes applied. If there are any other tweaks on top of this we may want before 1.12, we can use follow-up PRs.
